src/controller/CHIPDeviceController.cpp

```diff
@@ -990,12 +1034,87 @@ void DeviceCommissioner::OnDeviceAttestationInformationVerification(void * conte
                      static_cast<uint16_t>(result));
         // Go look at AttestationVerificationResult enum in src/credentials/attestation_verifier/DeviceAttestationVerifier.h to
         // understand the errors.
-        commissioner->CommissioningStageComplete(CHIP_ERROR_INTERNAL, report);
+
+        auto params = commissioner->mDefaultCommissioner->GetCommissioningParameters();
+        Credentials::DeviceAttestationDelegate * deviceAttestationDelegate = params.GetDeviceAttestationDelegate();
+
+        // If a device attestation status delegate is installed, delegate handling of failure to the client and let them
+        // decide on whether to proceed further or not.
+        if (deviceAttestationDelegate)
+        {
+            commissioner->ExtendArmFailSafeForFailedDeviceAttestation(result);
+        }
+        else
+        {
+            commissioner->CommissioningStageComplete(CHIP_ERROR_INTERNAL, report);
+        }
+    }
+    else
+    {
+        ChipLogProgress(Controller, "Successfully validated 'Attestation Information' command received from the device.");
+        commissioner->CommissioningStageComplete(CHIP_NO_ERROR);
+    }
+}
+
+void DeviceCommissioner::OnArmFailSafeExtendedForFailedDeviceAttestation(
+    void * context, const GeneralCommissioning::Commands::ArmFailSafeResponse::DecodableType & data)
+{
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+
+    if (!commissioner->mDeviceBeingCommissioned)
+    {
         return;
     }
 
-    ChipLogProgress(Controller, "Successfully validated 'Attestation Information' command received from the device.");
-    commissioner->CommissioningStageComplete(CHIP_NO_ERROR);
+    auto params = commissioner->mDefaultCommissioner->GetCommissioningParameters();
+    Credentials::DeviceAttestationDelegate * deviceAttestationDelegate = params.GetDeviceAttestationDelegate();
+    if (deviceAttestationDelegate)
+    {
+        ChipLogProgress(Controller, "Device attestation failed, delegating error handling to client");
+        NodeId remoteNodeId = commissioner->mDeviceBeingCommissioned->GetDeviceId();
+        deviceAttestationDelegate->OnDeviceAttestionFailed(commissioner, remoteNodeId, commissioner->mAttestationResult);
```

Comment:
I'm a little bit surprised that this behavior isn't just hidden away inside the `DeviceAttestationVerifier`.  Is there a reason why the commissioner needs to see the attestation result at all?  Why can't `DeviceAttestationVerifier` just tell the commissioner what to do directly through the `OnAttestationInformationVerification` callback interface?

Reply:
I believe this is to allow commissioners to allow failed attention to complete

Reply:
This is about requesting a client decide what to do when the attestation fails. This allows for the darwin framework to request input from the user and continue with commissioning even if an error is encountered.

Reply:
I understand the intent.  But the `DeviceAttestationVerifier` interface already provides such an opportunity.  Why is the attestation result coming into the common sdk commissioning code just to be sent back out to another implementation-specific delegate?  This design has:

```
Commissioner -> DeviceAttestationVerifier: request attestation verification from implementation delegate
Commissioner <- DeviceAttestationVerifier: provide attestation verification result to sdk
Commissioner -> DeviceAttestationDelegate: provide result to delegate, request next action
Commissioner <- DeviceAttestationDelegate: prescribe next action to commissioner
```

Why are we going in and out of the common sdk code twice?  What do we gain by having the commissioner shuttle data between two delegates when we can just have one?

Reply:
@msandstedt the device attestation verifier does not have the information it needs to extend the fail-safe, right?  Unless it has out-of-band knowledge of the DeviceCommissioner _and_ can get the device being commissioned from it...

Reply:
> @msandstedt the device attestation verifier does not have the information it needs to extend the fail-safe, right? Unless it has out-of-band knowledge of the DeviceCommissioner and can get the device being commissioned from it...

We the implementers will implement the device attestation verifier, so I would argue that if we need access to the device and the commissioner, then we can give ourselves this from within the verifier.  Isn't this the point of injecting an arbitrary object that implements the verifier interface?  It can have any context we need.  Also, attestation may be verified off-node.  The verifier may need to extend the failsafe anyway.  But I can see how this would be a more obvious implementation.

> given the need to change the fail-safe timer due to user input

I would just point out that this still may not work very well with a UI.  The amount of time for a user selection may be indeterminate.  It's likely the injected ecosystem object would still need to further manage the failsafe by periodically kicking it.

I think my critique is that the sdk can't be designed in a way that predicts all possible UI flows.  So why are we trying when we have a contiguous sequence of 'stuff' here that must all be managed by the app anyway.  If it all must be managed by the app anyway, and all happens in contiguous succession, why do we need to inject multiple delegates?

However, nothing here hinders implementations from doing what they need to do, so I suppose it is OK.

Reply:
> We the implementers will implement the device attestation verifier

In that case this PR is not really relevant.  This PR is mostly relevant if using the default attestation verifier....

Reply:
Ah, you have identified the disconnect then.  If there is a plausible path where the default verifier can actually talk to the DCL and have the current, valid set of PAAs, then I can see how people could want to use that, but then would need this.

Thanks.